### PR TITLE
CA-353011: Clean up certificates from unknown hosts

### PR DESCRIPTION
**Pull request by psafont**
Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

### CODE REVIEW

File: ocaml/xapi/db_gc_util.ml [hunk 1]
```diff
@@ -222,6 +222,23 @@ let gc_host_metrics ~__context =
...
+  let all_certificates = Db.Certificate.get_all_records ~__context in
+  let get_certificates =
+    List.map (fun self -> Db.Host.get_certificates ~__context ~self)
+  in
+  let host_certificates =
+    get_certificates (Db.Host.get_all ~__context) |> List.flatten
+  in
+  List.iter
+    (fun (cert, record) ->
+      if
```
Comment (lindig) [inline ocaml/xapi/db_gc_util.ml:235]: I would use `List.filter` followed by `List.iter` - it's not more efficient but I think it fits the paradigm and is readable.

Comment (lindig) [reply ocaml/xapi/db_gc_util.ml:235]: I think a comment about the intention would be also good: which certs you want to remote

Comment (lindig) [reply ocaml/xapi/db_gc_util.ml:235]: not (A || B) == (not A) && (not B) - would the  latter logic be more readable?

**Comment by robhoes** [comment]
Why not use `gc_connector`?

File: ocaml/xapi/db_gc_util.ml [hunk 2]
```diff
@@ -222,6 +222,23 @@ let gc_host_metrics ~__context =
...
     all_host_metrics
 
+let gc_certificates ~__context =
+  let all_certificates = Db.Certificate.get_all_records ~__context in
+  let get_certificates =
+    List.map (fun self -> Db.Host.get_certificates ~__context ~self)
+  in
+  let host_certificates =
+    get_certificates (Db.Host.get_all ~__context) |> List.flatten
+  in
```
Comment (lippirk) [inline ocaml/xapi/db_gc_util.ml:232]: you can use List.concat_map, I'd also just do it in one line

**Review by lippirk — APPROVED** [review]

**Comment by psafont** [comment]
As I understood it, `gc_connector` is used for classes that represent a many to many relationship. Certificates is does not fit that description and it must be handled separately.

File: ocaml/xapi/db_gc_util.ml (hunk 1 above)
Comment (psafont) [reply ocaml/xapi/db_gc_util.ml:235]: I did the && logic before, not sure which one is less confusing. I'll add a comment to state we want to remove the certificates that have a host associates (`host and `internal_host) with an invalid host reference.

File: ocaml/xapi/db_gc_util.ml (hunk 2 above)
Comment (psafont) [reply ocaml/xapi/db_gc_util.ml:232]: understand the eagerness, this needs to be backported to stockholm :p

**Review by lindig — APPROVED** [review]

File: ocaml/xapi/db_gc_util.ml (hunk 2 above)
Comment (psafont) [reply ocaml/xapi/db_gc_util.ml]: I'll change it in the backport ;p

**Comment by lindig** [comment]
Undraft and we merge it?

**Pull request merged**